### ISSUE TITLE
Sudden memory increase in ETCD members

### ISSUE THREAD

**Issue by armstrongli** [coreos/etcd#7964]
We got into sudden memory increase on ETCD members in k8s cluster, and the ETCD pods got OOM killed.

I did some investigation on the members, and found that the member's DISK(Network Volume provided by SolidFire) was a little slow, and the Memory of the member kept increasing until OOM killed. 
Then I switch the volume from network volume to local disk, and the OOM things gone.

Is it expected?

<img width="489" alt="screen shot 2017-05-22 at 10 45 13 pm" src="https://cloud.githubusercontent.com/assets/8025018/26314555/f29f7bf2-3f40-11e7-865a-07c65b3875fe.png">

**Comment by armstrongli** [comment]
The ETCD server is on `v3.0.15`

**Comment by xiang90** [comment]
Can you please provide etcd server log starting at 19:30 to 23:00 for all your members in the cluster?

**Comment by armstrongli** [comment]
The log is UTC time. The screenshot is Beijing time.
[etcd.log.2017.05.22.1.tar.gz](https://github.com/coreos/etcd/files/1026368/etcd.log.2017.05.22.1.tar.gz)

**Comment by armstrongli** [comment]
@xiang90 I think this can be the root solution: https://github.com/coreos/etcd/issues/7981#issuecomment-303795997

**Comment by xiang90** [comment]
There are quite a few lines on the graph. What are they? 

And I can see interleave logs. Were there two etcd concurrently writing into the same log file? It is very hard to understand what was going on from the log. 

But, yea, there are tons of warnings in your log. You need to make sure the warning rate is 0 most of the time. Or your etcd cluster will never be happy.

**Comment by armstrongli** [comment]
> Were there two etcd concurrently writing into the same log file? 

No. There's only one etcd running in the Node, and the etcd container is started by kubelet.

> there are tons of warnings in your log

The warnings are caused by disk sync I guess. And we have the local SSD and give ETCD enough resources. 
Do you have some recommendations to reduce the warnings or ease it out?

**Comment by armstrongli** [comment]
@xiang90 I use local SSD and start a whole new ETCD cluster  without any load , there're still bunch of warnings about applying entries.

```
/var/log # etcdctl --endpoints [$(echo $member_urls | sed -e "s/ /,/g")] --cacert=/etc/ssl/kubernetes/ca.crt endpoint status -w table;
+----------------------------------------------+------------------+---------+---------+-----------+-----------+------------+
|                   ENDPOINT                   |        ID        | VERSION | DB SIZE | IS LEADER | RAFT TERM | RAFT INDEX |
+----------------------------------------------+------------------+---------+---------+-----------+-----------+------------+
| https://tess-node-xq2sm-5915.51.tess.io:4001 | 5902a07919e43cdf | 3.0.15  | 25 kB   | true      |         4 |     108016 |
| https://tess-node-2o9hu-4886.51.tess.io:4001 | b620b4c395187fad | 3.0.15  | 25 kB   | false     |         4 |     108016 |
| https://tess-node-873t0-8911.51.tess.io:4001 | c2cb0e006c421dd3 | 3.0.15  | 25 kB   | false     |         4 |     108016 |
| https://tess-node-vh8tr-6888.51.tess.io:4001 | d49c2ef31cf07365 | 3.0.15  | 25 kB   | false     |         4 |     108016 |
| https://tess-node-ba93c-3871.51.tess.io:4001 | df3a73af6b1c2179 | 3.0.15  | 25 kB   | false     |         4 |     108017 |
+----------------------------------------------+------------------+---------+---------+-----------+-----------+------------+
/var/log # tail etcd.log
2017-05-26 04:48:07.465530 W | etcdserver: apply entries took too long [70.226991ms for 1 entries]
2017-05-26 04:48:07.465587 W | etcdserver: avoid queries with large range/delete range!
2017-05-26 04:51:10.465522 W | etcdserver: apply entries took too long [69.872088ms for 1 entries]
2017-05-26 04:51:10.465573 W | etcdserver: avoid queries with large range/delete range!
2017-05-26 04:51:11.465581 W | etcdserver: apply entries took too long [70.156267ms for 1 entries]
2017-05-26 04:51:11.465624 W | etcdserver: avoid queries with large range/delete range!
2017-05-26 05:01:11.965522 W | etcdserver: apply entries took too long [70.038266ms for 1 entries]
2017-05-26 05:01:11.965589 W | etcdserver: avoid queries with large range/delete range!
2017-05-26 05:02:46.965542 W | etcdserver: apply entries took too long [28.357591ms for 1 entries]
2017-05-26 05:02:46.965601 W | etcdserver: avoid queries with large range/delete range!
/var/log #
```

**Comment by xiang90** [comment]
> 108016

why raft index is so high?

does any other application share the same SSD? if yes, move them away. or there is an issue with your SSD. probably ask your hardware guys to get it fixed.

**Comment by armstrongli** [comment]
> does any other application share the same SSD?

No. There're only another 2 applications running on the same node. One is prometheus which is using network volume. The other one is grafana dashboard which is not using any disk IO.

**Comment by xiang90** [comment]
@armstrongli 

Try to kill your another two applications. If the problem still exists, then you probably need to ask your hardware guy about what is going on. if etcd under no loads, apply should always be one or zero simple disk hit.

**Comment by armstrongli** [comment]
I'll check the hardware level things about the performance issue.

**Comment by zbindenren** [comment]
Hi

We have a similar problem with version 3.1.3.
![memory](https://cloud.githubusercontent.com/assets/1607088/26621175/1cdd5000-45e6-11e7-95b0-ca715a50dde4.png)

![leader_elections](https://cloud.githubusercontent.com/assets/1607088/26621192/2acd65d8-45e6-11e7-867b-29d8b261fe1f.png)

It is always the leader that consumes the memory.

We have 3 Nodes with around 1200 clients connected.

**Comment by xiang90** [comment]
@zbindenren I believe your issues should be fixed if you upgrade to 3.1.5+. There is a bug in the etcd release > 3.1 but < 3.1.5.

@armstrongli 

I am going to close this issue since it is inactive and we do not have enough information to help you.

**Comment by armstrongli** [comment]
@xiang90 Thanks. And I'll upgrade our cluster to v3.1.9 .

**Comment by zbindenren** [comment]
Problem is fixed with 3.1.9
